### ISSUE TITLE
read_csv with wildcards fails on different column orders

### ISSUE THREAD

**Issue by stumitchell** [dask/dask#2053]
Hi I finally tracked down this issue after alot of work but if you have 2 (or more) csv files with headers, but the column order of the files is different, read_csv will read the second file as if the columns where in the same order.

I did not know that the order of these columns would be different but `pd.Dataframe.to_csv()` does not guarantee an ordering.

take for example the following 2 files.

    ==> mean-2016.12.11.csv <==
    d,demo,s,market,qtr_hr,ue,000s,w,np,tarp
    2016-12-11,TTLPPL,0,0201,0,4998799.939999982,6.332151999999998,6332.151999999998,5,0.12667344314643691
    2016-12-11,TTLPPL,0,0201,1,4998799.939999982,3.9752799999999993,3975.2799999999993,2,0.07952468687914752
    2016-12-11,TTLPPL,0,0201,2,4998799.939999982,3.2186719999999993,3218.6719999999996,2,0.0643888941072527

    ==> mean-2016.12.18.csv <==
    d,demo,s,market,qtr_hr,w,ue,np,000s,tarp
    2016-12-18,TTLPPL,0,0201,0,1743.1853333333338,4998800.000000004,2,1.7431853333333331,0.034872075964898205
    2016-12-18,TTLPPL,0,0201,1,1952.8600000000004,4998800.000000004,3,1.9528599999999998,0.039066575978234735
    2016-12-18,TTLPPL,0,0201,2,1654.542666666667,4998800.000000004,2,1.6545426666666665,0.033098797044624005

you get the following

    >>> df = dd.read_csv('../data/mean-2016.12.18.csv')
    >>> df.compute().ix[0]
    d         2016-12-18
    demo          TTLPPL
    s                  0
    market           201
    qtr_hr             0
    w            1743.19
    ue        4.9988e+06
    np                 2
    000s         1.74319
    tarp       0.0348721
    Name: 0, dtype: object

in contrast

    >>> df = dd.read_csv('../data/mean-2016.12.1*.csv')
    >>> df.compute().ix[0]

    d	demo	s	market	qtr_hr	ue	000s	w	np	tarp
    0	2016-12-11	TTLPPL	0	201	0	4.998800e+06	6.332152e+00	6332.152	5.000000	0.126673
    0	2016-12-18	TTLPPL	0	201	0	1.743185e+03	4.998800e+06	2.000	1.743185	0.034872

**Comment by martindurant** [comment]
Correct, dask's dataframe.read_csv is designed so that the header from the first file can be applied to all, or that you supply a header that can be applied to all. This is necessary when splitting large files into chunks, since the header is applied to each chunk.

I do not immediately see a combination of parameters that gets around this (perhaps someone else knows, but if you do not need to split up your files, you could do something like this:
```
df = dd.from_delayed([dask.delayed(pd.read_csv)(f) for f in filelist])
```
(the dataframe pieces will still have their internal ordering of columns differently, but that won't matter when almost everything works by selecting columns by label)

**Comment by stumitchell** [comment]
yeah I got it around it like this 

   df = dd.concat([dd.read_csv(f) for f in filelist])

but that means I can't use wildcards anymore :-(

but If it can't be fixed it should produce an error (or at least a warning).

**Comment by mrocklin** [comment]
> but If it can't be fixed it should produce an error (or at least a warning).

I agree that it should do this.  There are literally dozens of cases with malformed CSV where it would be useful to provide informative errors.  Unfortunately there is no way to make the directory-of-csv format entirely reliable without looking through all of the data ahead of time.  There is a necessary trade-off here between "get going quickly" and "thoroughly check through all of the data for inconsistencies".

If you have thoughts on how to improve the current internal design of `dd.read_csv` and the time to implement those changes then contributions would be most welcome.

**Comment by ofirshifman** [comment]
long time since it was opened, but I still wonder what is the best practice for this issue. please share from you reasults or join the conversation here:

https://stackoverflow.com/questions/68647921/using-dask-to-load-many-csv-files-with-different-columns

**Comment by jrbourbeau** [comment]
It looks like https://github.com/dask/dask/issues/2053#issuecomment-285060500 still reflects the current status
